### PR TITLE
feat(edge): calculate endpoints of edge

### PR DESCRIPTION
**Pull request by yvonneyx**
#### 计算边上端点的实际连接位置

- 两端节点未指定 `anchorOptions`
    - _(如左图1)_ 计算从一个节点的几何中心到另一个节点几何中心的连线，并找到这条连线与节点边界的交点。这两个交点作为边的起点与终点。
- 两端节点指定 `anchorOptions`
    - _(如左图2)_ 若边上明确指定 `sourceAnchor`(anchor key) 和 `targetAnchor` 则直接作为边的起点与终点。
    - _(如左图3、4、5)_ 若边上未指定，则自动选择两个节点上距离最近的锚点作为边的起点与终点。
- 一端节点指定，一端未指定
    - _(如右图1、2、3、4)_ 在这种情况下，对于指定了 `anchorOptions` 的一端，将使用其指定的锚点作为连接点。对于未指定 `anchorOptions` 的一端，则采取类似于两端都未指定 `anchorOptions` 的处理方式：计算该端节点的几何中心到另一端指定的锚点（或几何中心，如果另一端也未指定锚点）的连线，并找到该连线与节点边界的交点，用作连接线的起点或终点

------

#### Calculates anchor positions of the endpoints on an edge

- Both end nodes not specified `anchorOptions`
    - _(as in Figure 1 on the left)_ Calculates a line from the geometric center of one node to the geometric center of another node and finds the intersection of this line with the node boundary. These two intersections serve as the start and end points of the edge.
- The two end nodes are specified `anchorOptions`
    - _(e.g. Figure 2 on the left)_ If `sourceAnchor` (anchor key) and `targetAnchor` are explicitly specified on the edge, they are used directly as the start and end points of the edge.
    - _(As shown in the left figure 3, 4, 5)_ If the edge is not specified, the closest anchor point of the two nodes is automatically selected as the start and end point of the edge.
- One end node specified, one end unspecified
    - _(e.g. 1, 2, 3, 4 on the right)_ In this case, for the end where `anchorOptions` is specified, its specified anchor point will be used as the connection point. For the end that does not specify `anchorOptions`, it is treated in a similar way as if neither end specified `anchorOptions`: the line from the geometric center of the node at that end to the specified anchor (or geometric center, if the other end does not specify an anchor either) is computed, and the intersection of the line with the boundary of the node is found and used as the start or end point of the line.

endpoints

### CODE REVIEW

File: packages/g6/src/elements/nodes/base-node.ts [hunk 1]
```diff
@@ -107,6 +109,19 @@ export abstract class BaseNode<KT extends object, KS> extends BaseShape<BaseNode
     });
   }
 
+  public getAnchors(): Record<string, GCircle> {
+    return extractWithPrefix(this.shapeMap, 'anchor-');
```
Comment (Aarebecca) [inline packages/g6/src/elements/nodes/base-node.ts:113]: 这个是不是和 subStyleProps 的逻辑重合了

File: packages/g6/src/utils/point.ts [hunk 2]
```diff
@@ -37,22 +50,150 @@ export function isVertical(p1: Point, p2: Point): boolean {
...
+ *
+ * <en/> Gets the intersection point between the line from the center of an ellipse to a given point and the ellipse's edge
+ * @param p - <zh/> 从椭圆中心到椭圆边缘的连线的外部点 | <en/> The point outside the ellipse from which the line to the ellipse's center is drawn
+ * The point outside the ellipse from which the line to the ellipse's center is drawn.
+ * @param bbox - <zh/> 椭圆包围盒 | <en/> the bounding box of the ellipse
+ * @returns <zh/> 交点 | <en/> intersection
+ */
+export function getEllipseIntersectPoint(p: Point, bbox: AABB): Point {
+  const rx = getBBoxWidth(bbox) / 2;
+  const ry = getBBoxWidth(bbox) / 2;
```
Comment (Aarebecca) [inline packages/g6/src/utils/point.ts:167]: 下面这个是 `getBBoxHeight` ?

File: packages/g6/src/utils/point.ts [hunk 3]
```diff
@@ -37,22 +50,150 @@ export function isVertical(p1: Point, p2: Point): boolean {
...
+ * <zh/> 判断两条线段是否平行
+ *
+ * <en/> Judge whether two line segments are parallel
+ * @param p1 - <zh/> 第一条线段的起点 | <en/> the start point of the first line segment
+ * @param p2 - <zh/> 第一条线段的终点 | <en/> the end point of the first line segment
+ * @param p3 - <zh/> 第二条线段的起点 | <en/> the start point of the second line segment
+ * @param p4 - <zh/> 第二条线段的终点 | <en/> the end point of the second line segment
+ * @returns <zh/> 是否平行 | <en/> whether parallel or not
+ */
+export function isLinesParallel(p1: Point, p2: Point, p3: Point, p4: Point): boolean {
```
Comment (Aarebecca) [inline packages/g6/src/utils/point.ts:78]: 这里用两个数组来表示更容易理解些？ `isLinesParallel(l1: [Point, Point], l2: [Point, Point])`

File: packages/g6/src/elements/nodes/base-node.ts (hunk 1 above)
Comment (yvonneyx) [reply packages/g6/src/elements/nodes/base-node.ts:113]: - `subStyleProps('haloFill', 'halo) => 'fill'` 中有一部分定制逻辑，这边不太适用
- 命名统一，调整成 `subObject`

File: packages/g6/src/utils/point.ts [hunk 4]
```diff
@@ -37,22 +50,150 @@ export function isVertical(p1: Point, p2: Point): boolean {
...
+ * <zh/> 获取两条线段的交点
+ *
+ * <en/> Get the intersection of two line segments
+ * @param p1 - <zh/> 第一条线段的起点 | <en/> the start point of the first line segment
+ * @param p2 - <zh/> 第一条线段的终点 | <en/> the end point of the first line segment
+ * @param p3 - <zh/> 第二条线段的起点 | <en/> the start point of the second line segment
+ * @param p4 - <zh/> 第二条线段的终点 | <en/> the end point of the second line segment
+ * @returns <zh/> 交点 | <en/> intersection
+ */
+export function getLinesIntersection(p1: Point, p2: Point, p3: Point, p4: Point): Point | undefined {
```
Comment (Aarebecca) [inline packages/g6/src/utils/point.ts:94]: 这里一起改下吧～

**Review by Aarebecca — APPROVED** [review]

**Pull request merged**